### PR TITLE
Support hostname waypoints

### PR DESCRIPTION
**Pull request by howardjohn**
Fixes https://github.com/istio/istio/issues/52318

### CODE REVIEW

**Comment by ilrudie** [comment]
https://github.com/istio/ztunnel/pull/1246 was merged into 1.23 release so we should be go for merging the istiod changes

File: pilot/pkg/config/kube/gateway/conversion.go [hunk 1]
```diff
@@ -2244,10 +2244,11 @@ func reportGatewayStatus(
...
 			}
 			// There are no external addresses, so report the internal ones
-			// TODO: should we always report both?
-			if wantAddressType == k8s.IPAddressType {
+			// This can be IP, Hostname, or both (indicated by empty wantAddressType)
+			if wantAddressType != k8s.HostnameAddressType {
 				addressesToReport = internalIP
-			} else {
+			}
+			if wantAddressType != k8s.IPAddressType {
```
Comment (hzxuzhonghu) [inline pilot/pkg/config/kube/gateway/conversion.go:2251]: why not use else here

File: pilot/pkg/config/kube/gateway/deploymentcontroller.go [hunk 2]
```diff
@@ -168,7 +168,7 @@ func getClassInfos() map[gateway.GatewayController]classInfo {
 			templates:          "waypoint",
 			disableNameSuffix:  true,
 			defaultServiceType: corev1.ServiceTypeClusterIP,
-			addressType:        gateway.IPAddressType,
+			addressType:        "", // Report both
```
Comment (hzxuzhonghu) [inline pilot/pkg/config/kube/gateway/deploymentcontroller.go:171]: Please comment why it need both kind of address reported

File: pilot/pkg/serviceregistry/kube/controller/ambient/helpers.go [hunk 3]
```diff
@@ -78,16 +78,8 @@ func byteIPToAddr(b []byte) netip.Addr {
 }
 
 func (a *index) getWaypointAddress(w *Waypoint) *workloadapi.GatewayAddress {
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/ambient/helpers.go:80]: The func can be removed now

File: pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go [hunk 4]
```diff
@@ -357,12 +356,40 @@ func makeAllowedRoutes(gateway *v1beta1.Gateway) WaypointSelector {
...
 
-func getGatewayAddrs(gw *v1beta1.Gateway) []netip.Addr {
-	// Currently, we only look at one address. Probably this should be made more robust
-	ip, err := netip.ParseAddr(gw.Status.Addresses[0].Value)
-	if err == nil {
-		return []netip.Addr{ip}
+func (a *index) getGatewayAddress(gw *v1beta1.Gateway) *workloadapi.GatewayAddress {
+	for _, addr := range gw.Status.Addresses {
+		if addr.Type != nil && *addr.Type == v1beta1.HostnameAddressType {
+			// Prefer hostname from status, if we can find it.
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go:362]: why do you prefer hostname address

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/waypoints.go:362]: Added a comment

File: pilot/pkg/config/kube/gateway/conversion.go (hunk 1 above)
Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/conversion.go:2251]: `else` doesn't handle setting both

**Review by dhawton — APPROVED** [review]

**Pull request merged**